### PR TITLE
Support long sequence in attention

### PR DESCRIPTION
**Pull request by zhanghuanrong**
Support long sequence in attention operator for (1) raw mask of 2/3/4-D, (2) no mask.
Set longer greedy search max length.

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/transformers/greedy_search_parameters.cc [hunk 1]
```diff
@@ -6,7 +6,7 @@ namespace onnxruntime {
 namespace contrib {
 namespace transformers {
 
-constexpr int kMaxSequenceLength = 4096;
+constexpr int kMaxSequenceLength = 9000; // support more than 8K
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/greedy_search_parameters.cc:9]: shall we use 8192 or 16384?

File: onnxruntime/contrib_ops/cuda/bert/attention_softmax.h [hunk 2]
```diff
@@ -169,6 +169,189 @@ __device__ inline void SoftmaxSmall(const int all_sequence_length,
...
+    } else {
+      end = min(valid_end, end_unid);
+    }
+  }
+
+  // Input dimension is BxNxSxS*; blockIdx.y is batch index b; gridDim.x=N*S;  blockIdx.x is index within N*S;
+  const int offset = (blockIdx.y * gridDim.x + blockIdx.x) * all_sequence_length;
+
+  float thread_data_max = -CUDART_INF_F;
+  for (int thx = threadIdx.x; thx < all_sequence_length; thx += TPB) {
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/attention_softmax.h:210]: >thx [](http://example.com/codeflow?start=11&length=3)

nit: a better name, like seq_idx

File: onnxruntime/contrib_ops/cuda/bert/attention_softmax.h [hunk 3]
```diff
@@ -169,6 +169,189 @@ __device__ inline void SoftmaxSmall(const int all_sequence_length,
...
+  const auto max = BlockReduce(tmp_storage).Reduce(thread_data_max, cub::Max(), end);
+
+  // Store max value
+  if (threadIdx.x == 0) {
+    max_block = max;
+  }
+  __syncthreads();
+
+  float thread_data_exp(0.f);
+  for (int thx = threadIdx.x; thx < all_sequence_length; thx += TPB) {
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/attention_softmax.h:233]: >thx [](http://example.com/codeflow?start=11&length=3)

nit: a better name

File: onnxruntime/contrib_ops/cuda/bert/attention_softmax.h [hunk 4]
```diff
@@ -315,7 +498,10 @@ Status ComputeSoftmax(cudaStream_t stream, const int all_sequence_length, const
     SoftmaxKernel<T, blockSize><<<grid, blockSize, 0, stream>>>(
         all_sequence_length, sequence_length, add_before_softmax, input, output);
   } else {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Attention CUDA operator does not support total sequence length > 1024.");
+    const int blockSize = 1024;
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/attention_softmax.h:501]: >blockSize [](http://example.com/codeflow?start=14&length=9)

a smaller blocksize will be better considering you are using BlockReduce.

File: onnxruntime/contrib_ops/cuda/bert/attention_softmax.h [hunk 5]
```diff
@@ -169,6 +169,189 @@ __device__ inline void SoftmaxSmall(const int all_sequence_length,
...
+
+  __shared__ float sum_reverse_block;
+  __shared__ float max_block;
+
+  // Update end position for unidirectional.
+  int end = valid_end;
+  int end_unid = -1;
+  if (is_unidirectional) {
+    end_unid = all_sequence_length - sequence_length + (blockIdx.x % sequence_length) + 1;
+    if (end_unid <= valid_start) {
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/attention_softmax.h:194]: in which cast this will happen?

**Comment by yufenglee** [comment]
The implementation looks good to me overall. However, there is an algo that can implement softmax with only one reduce. For large vector size, online can get much better performance.
Here is implementation example and paper:
[https://github.com/NVIDIA/online-softmax/blob/f4c607fa358789d988b1ff346393c483b457e230/online_softmax_benchmark.cu#L169](https://github.com/NVIDIA/online-softmax/blob/f4c607fa358789d988b1ff346393c483b457e230/online_softmax_benchmark.cu#L169)
[https://arxiv.org/pdf/1805.02867.pdf](https://arxiv.org/pdf/1805.02867.pdf)

**Review by tianleiwu — COMMENTED** [review]
Could you also compare with TensorRT's softmax:
https://github.com/NVIDIA/TensorRT/blob/1d6bf36034bb07fc88d924ebf5d34d358298e545/plugin/bertQKVToContextPlugin/qkvToContext.cu#L37

File: onnxruntime/contrib_ops/cuda/bert/attention_softmax.h (hunk 3 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cuda/bert/attention_softmax.h:233]: Fix.

File: onnxruntime/contrib_ops/cuda/bert/attention_softmax.h (hunk 5 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cuda/bert/attention_softmax.h:194]: Just follow original code.

**Comment by zhanghuanrong** [comment]
This change does not target to best perf. It extent existing logic and try to get same float result. Single reduce path is definitely faster, but float computation is different. We could leave it later. TRT's implementation need to branch on more sequence length to do the pragma unroll, it could be faster on specific sequence length, yet still need general it we want longer sequence support in future. Just Add base general here. And perf tuning could be done on specific use case.

**Comment by yufenglee** [comment]
> This change does not target to best perf. It extent existing logic and try to get same float result. Single reduce path is definitely faster, but float computation is different. We could leave it later. TRT's implementation need to branch on more sequence length to do the pragma unroll, it could be faster on specific sequence length, yet still need general it we want longer sequence support in future. Just Add base general here. And perf tuning could be done on specific use case.

Which float computation is different? With the online softmax, we only need 1 kernel instead of multiple.

**Review by yufenglee — APPROVED** [review]

File: onnxruntime/contrib_ops/cuda/bert/attention_softmax.h [hunk 6]
```diff
@@ -169,6 +169,189 @@ __device__ inline void SoftmaxSmall(const int all_sequence_length,
...
+      ;
```
Comment (yuslepukhin) [inline onnxruntime/contrib_ops/cuda/bert/attention_softmax.h:195]: >   ; [](http://example.com/codeflow?start=3&length=4)

Dead code?

**Pull request merged**